From bc1d6d1529a2eaeedaee1a02a3ac77a687d89977 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Tue, 25 Jun 2024 18:07:39 +0200
Subject: [PATCH] Fix CoinbaseMaturity upgrade (#12)

It was erroneously assumed that the upgrade height would be based on the height where the coin is *spent*, not where it's *mined*, which is how it actually works in Dogecoin.

This basically simply offsets the upgrade.

diff --git a/src/consensus/tx_verify.cpp b/src/consensus/tx_verify.cpp
index ea9c6cb7d42f96cd5952626e83e8443a670bd249..e9d7f6510cecb81f668b494fc44563816a7d5919 100644
--- a/src/consensus/tx_verify.cpp
+++ b/src/consensus/tx_verify.cpp
@@ -203,9 +203,13 @@ bool CheckTxInputs(const CTransaction &tx, TxValidationState &state,
         const Coin &coin = inputs.AccessCoin(prevout);
         assert(!coin.IsSpent());
 
+        // Dogecoin: Coinbase maturity depends on the coin's height
+        const int32_t minCoinbaseMaturity =
+            CoinbaseMaturity(params, coin.GetHeight());
+
         // If prev is coinbase, check that it's matured
-        if (coin.IsCoinBase() && nSpendHeight - (int)coin.GetHeight() <
-                                     CoinbaseMaturity(params, nSpendHeight)) {
+        if (coin.IsCoinBase() &&
+            nSpendHeight - (int)coin.GetHeight() < minCoinbaseMaturity) {
             return state.Invalid(
                 TxValidationResult::TX_PREMATURE_SPEND,
                 "bad-txns-premature-spend-of-coinbase",
diff --git a/src/qt/transactiondesc.cpp b/src/qt/transactiondesc.cpp
index 1a6003f7aa19aeddb4c04ad1adce3d73eee38812..b17b3e7fd9758b024b80eb54a32e55d28ec2ad18 100644
--- a/src/qt/transactiondesc.cpp
+++ b/src/qt/transactiondesc.cpp
@@ -359,7 +359,7 @@ QString TransactionDesc::toHTML(interfaces::Node &node,
     if (wtx.is_coinbase) {
         quint32 numBlocksToMaturity =
             CoinbaseMaturity(wallet.getChainParams().GetConsensus(),
-                             wallet.wallet()->GetLastBlockHeight()) +
+                             status.block_height) +
             1;
         strHTML +=
             "<br>" +
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index b278e33d815bbaadba2aaf555dd07aa1bf69be9e..e3cca5eaef2d4fe8ebd517d5536dd0fa4b3edf0a 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -3263,8 +3263,8 @@ int CWallet::GetTxBlocksToMaturity(const CWalletTx &wtx) const {
     if (!wtx.IsCoinBase()) {
         return 0;
     }
-    const int32_t coinbaseMaturity =
-        CoinbaseMaturity(GetChainParams().GetConsensus(), GetLastBlockHeight());
+    const int32_t coinbaseMaturity = CoinbaseMaturity(
+        GetChainParams().GetConsensus(), wtx.m_confirm.block_height);
     int chain_depth = GetTxDepthInMainChain(wtx);
     // coinbase tx should not be conflicted
     assert(chain_depth >= 0);
diff --git a/test/functional/feature_upgrade_coinbase_maturity.py b/test/functional/feature_upgrade_coinbase_maturity.py
index dbd495f1febdbfdde36eef4cf9527a3a24187e0c..83eb6f049e79efaf960bf2b40505a6e816188554 100644
--- a/test/functional/feature_upgrade_coinbase_maturity.py
+++ b/test/functional/feature_upgrade_coinbase_maturity.py
@@ -1,7 +1,7 @@
 # Copyright (c) 2024 The Bitcoin developers
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
-"""Test coinbase maturity upgrades."""
+"""Test Dogecoin's coinbase maturity upgrade."""
 
 from test_framework.address import (
     ADDRESS_ECREG_P2SH_OP_TRUE,
@@ -61,70 +61,54 @@ class UpgradeCoinbaseMaturityTest(BitcoinTestFramework):
         minedhash = self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)[0]
         assert_equal(node.getrawtransaction(tx.hash, 2)["blockhash"], minedhash)
 
-        # Generate blocks until we're 241 blocks before upgrade height
+        # Generate blocks until we're 2 blocks before upgrade height
         UPGRADE_HEIGHT = 1450
         OLD_MATURITY = 100
         DIGISHIELD_MATURITY = 240
         self.generatetoaddress(
             node,
-            UPGRADE_HEIGHT - DIGISHIELD_MATURITY - OLD_MATURITY - 2,
+            UPGRADE_HEIGHT - 103,
             ADDRESS_ECREG_UNSPENDABLE,
         )
-        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - DIGISHIELD_MATURITY - 1)
-
-        # Get us two coins with 239 and 240 maturity at upgrade
-        mature240hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        mature240txid = node.getblock(mature240hash)["tx"][0]
-        mature239hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        mature239txid = node.getblock(mature239hash)["tx"][0]
-        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - DIGISHIELD_MATURITY + 1)
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - 2)
 
-        self.generatetoaddress(
-            node, DIGISHIELD_MATURITY - OLD_MATURITY - 3, ADDRESS_ECREG_UNSPENDABLE
-        )
-        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - OLD_MATURITY - 2)
+        # Get us a coin one block before upgrade
+        preupgradehash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
+        preupgradetxid = node.getblock(preupgradehash)["tx"][0]
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - 1)
 
-        # Generate two coins with 99 and 100 maturity one block before upgrade
-        mature100hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        mature100txid = node.getblock(mature100hash)["tx"][0]
-        mature99hash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
-        mature99txid = node.getblock(mature99hash)["tx"][0]
-        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - OLD_MATURITY)
+        # Get us a coin after the upgrade
+        postupgradehash = self.generatetoaddress(node, 1, ADDRESS_ECREG_P2SH_OP_TRUE)[0]
+        postupgradetxid = node.getblock(postupgradehash)["tx"][0]
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT)
 
-        self.generatetoaddress(node, OLD_MATURITY - 2, ADDRESS_ECREG_UNSPENDABLE)
-        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - 2)
+        # After that many blocks, we had 9 halvings
+        coinvalue >>= 9
 
-        # After that many blocks, we had 8 halvings
-        coinvalue >>= 8
+        # Mature first coin to 98 (mempool thinks it's 99)
+        self.generatetoaddress(node, OLD_MATURITY - 3, ADDRESS_ECREG_UNSPENDABLE)
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT + OLD_MATURITY - 3)
 
-        # For the mempool, we are still 1 block before upgrade.
-        # We can spend the 100 mature coin fine, ...
-        spend100tx = CTransaction()
-        spend100tx.vin = [
-            CTxIn(COutPoint(int(mature100txid, 16), 0), SCRIPTSIG_OP_TRUE)
+        # The pre-upgrade coin still needs 1 extra block
+        spend_preupgradetx = CTransaction()
+        spend_preupgradetx.vin = [
+            CTxIn(COutPoint(int(preupgradetxid, 16), 0), SCRIPTSIG_OP_TRUE)
         ]
-        spend100tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
-        pad_tx(spend100tx)
-        node.sendrawtransaction(spend100tx.serialize().hex())
-
-        # ...but not the 99 mature coin
-        spend99tx = CTransaction()
-        spend99tx.vin = [CTxIn(COutPoint(int(mature99txid, 16), 0), SCRIPTSIG_OP_TRUE)]
-        spend99tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
-        pad_tx(spend99tx)
+        spend_preupgradetx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
+        pad_tx(spend_preupgradetx)
         assert_raises_rpc_error(
             -26,
             "bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 99",
             node.sendrawtransaction,
-            spend99tx.serialize().hex(),
+            spend_preupgradetx.serialize().hex(),
         )
 
-        # Can't mine the 99 mature coin spend, ...
+        # Can't mine the 99 mature coin spend
         block = create_block(
             int(node.getbestblockhash(), 16), create_coinbase(node.getblockcount() + 1)
         )
         block.nVersion = 4
-        block.vtx += [spend99tx]
+        block.vtx += [spend_preupgradetx]
         block.hashMerkleRoot = block.calc_merkle_root()
         block.solve()
         peer.send_blocks_and_test(
@@ -133,47 +117,44 @@ class UpgradeCoinbaseMaturityTest(BitcoinTestFramework):
             success=False,
             reject_reason="bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 99",
             expect_disconnect=True,
-            timeout=5,
         )
         peer = self.reconnect_p2p()
 
-        # ... but we can mine the 100 mature coin spend
-        # Also, this activates the upgrade
+        # After 1 extra block, pre-upgrade coin is mature
+        self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)
+        node.sendrawtransaction(spend_preupgradetx.serialize().hex())
+
+        # Make sure we actually mined the tx
         minedhash1 = self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)[0]
         assert_equal(
-            node.getrawtransaction(spend100tx.hash, 2)["blockhash"], minedhash1
+            node.getrawtransaction(spend_preupgradetx.hash, 2)["blockhash"], minedhash1
         )
-        assert_equal(node.getblockcount(), UPGRADE_HEIGHT - 1)
 
-        # For the mempool the upgrade is now activated, so we can't spend a 239 mature coin
-        spend239tx = CTransaction()
-        spend239tx.vin = [
-            CTxIn(COutPoint(int(mature239txid, 16), 0), SCRIPTSIG_OP_TRUE)
+        # Mature second coin to 238 (mempool thinks it's 239)
+        self.generatetoaddress(
+            node, DIGISHIELD_MATURITY - OLD_MATURITY - 1, ADDRESS_ECREG_UNSPENDABLE
+        )
+        assert_equal(node.getblockcount(), UPGRADE_HEIGHT + DIGISHIELD_MATURITY - 2)
+
+        spend_postupgradetx = CTransaction()
+        spend_postupgradetx.vin = [
+            CTxIn(COutPoint(int(postupgradetxid, 16), 0), SCRIPTSIG_OP_TRUE)
         ]
-        spend239tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
-        pad_tx(spend239tx)
+        spend_postupgradetx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
+        pad_tx(spend_postupgradetx)
         assert_raises_rpc_error(
             -26,
             "bad-txns-premature-spend-of-coinbase, tried to spend coinbase at depth 239",
             node.sendrawtransaction,
-            spend239tx.serialize().hex(),
+            spend_postupgradetx.serialize().hex(),
         )
 
-        # But, we can spend the 240 mature coin
-        spend240tx = CTransaction()
-        spend240tx.vin = [
-            CTxIn(COutPoint(int(mature240txid, 16), 0), SCRIPTSIG_OP_TRUE)
-        ]
-        spend240tx.vout = [CTxOut(coinvalue - 10000, SCRIPT_UNSPENDABLE)]
-        pad_tx(spend240tx)
-        node.sendrawtransaction(spend240tx.serialize().hex())
-
-        # Can't mine the 239 mature coin spend, ...
+        # Can't mine the 239 mature coin spend
         block = create_block(
             int(node.getbestblockhash(), 16), create_coinbase(node.getblockcount() + 1)
         )
         block.nVersion = 4
-        block.vtx += [spend239tx]
+        block.vtx += [spend_postupgradetx]
         block.hashMerkleRoot = block.calc_merkle_root()
         block.solve()
         peer.send_blocks_and_test(
@@ -185,12 +166,15 @@ class UpgradeCoinbaseMaturityTest(BitcoinTestFramework):
         )
         peer = self.reconnect_p2p()
 
-        # ... but we can mine the 240 mature coin spend
+        # After 1 extra block, post-upgrade coin is mature
+        self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)
+        node.sendrawtransaction(spend_postupgradetx.serialize().hex())
+
+        # Make sure we actually mined the tx
         minedhash2 = self.generatetoaddress(node, 1, ADDRESS_ECREG_UNSPENDABLE)[0]
         assert_equal(
-            node.getrawtransaction(spend240tx.hash, 2)["blockhash"], minedhash2
+            node.getrawtransaction(spend_postupgradetx.hash, 2)["blockhash"], minedhash2
         )
-        assert_equal(node.getblockcount(), UPGRADE_HEIGHT)
 
 
 if __name__ == "__main__":
